libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1370,17 +1376,37 @@ void ExpressionCompiler::appendBitOperatorCode(Token::Value _operator)
 	}
 }
 
-void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator)
+void ExpressionCompiler::appendShiftOperatorCode(Token::Value _operator, Type const& _leftType, Type const& _rightType)
 {
-	BOOST_THROW_EXCEPTION(InternalCompilerError() << errinfo_comment("Shift operators not yet implemented."));
+	// stack: rvalue lvalue
+
+	IntegerType const& leftType = dynamic_cast<IntegerType const&>(_leftType);
+	bool const c_leftSigned = leftType.isSigned();
+
+	// The RValue can be a RationalNumberType too.
+	bool c_rightSigned = false;
+	if (auto rightType = dynamic_cast<RationalNumberType const*>(&_rightType))
+		c_rightSigned = rightType->integerType()->isSigned();
+	else if (auto rightType = dynamic_cast<IntegerType const*>(&_rightType))
+		c_rightSigned = rightType->isSigned();
+
+	// shift with negative rvalue throws exception
+	if (c_rightSigned)
+	{
+		m_context << u256(0) << Instruction::DUP3 << Instruction::SLT;
+		m_context.appendConditionalJumpTo(m_context.errorTag());
+	}
```

Comment:
Shall we also throw when the right hand side is bigger than the width of the type of the left hand side?

Another possibility is to throw for signed overflows.  This makes `(int3) 1 << 2` throw, because 4 is not representable in `int3`.